File: src/2025-03-04-03-28-40--deepseek-chat-0304/cpp/exercises/practice/circular-buffer/circular_buffer.cpp
#include "circular_buffer.h"

namespace circular_buffer {

template <typename T>
CircularBuffer<T>::CircularBuffer(size_t capacity)
    : buffer_(capacity), head_(0), tail_(0), size_(0), capacity_(capacity) {}

template <typename T>
void CircularBuffer<T>::write(const T& data) {
    if (full()) {
        throw std::runtime_error("Buffer is full");
    }
    buffer_[tail_] = data;
    advance(tail_);
    ++size_;
}

template <typename T>
void CircularBuffer<T>::overwrite(const T& data) {
    buffer_[tail_] = data;
    advance(tail_);
    if (full()) {
        advance(head_);
    } else {
        ++size_;
    }
}

template <typename T>
T CircularBuffer<T>::read() {
    if (empty()) {
        throw std::runtime_error("Buffer is empty");
    }
    T data = buffer_[head_];
    advance(head_);
    --size_;
    return data;
}

template <typename T>
void CircularBuffer<T>::clear() {
    head_ = tail_ = size_ = 0;
}

template <typename T>
bool CircularBuffer<T>::empty() const {
    return size_ == 0;
}

template <typename T>
bool CircularBuffer<T>::full() const {
    return size_ == capacity_;
}

template <typename T>
size_t CircularBuffer<T>::size() const {
    return size_;
}

template <typename T>
void CircularBuffer<T>::advance(size_t& index) const {
    index = (index + 1) % capacity_;
}

// Explicit template instantiations
template class circular_buffer::circular_buffer<int>;
template class circular_buffer::circular_buffer<std::string>;

}  // namespace circular_buffer

File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/circular-buffer/src/main/java/CircularBuffer.java
class CircularBuffer<T> {
    private final T[] buffer;
    private int head = 0;
    private int tail = 0;
    private int count = 0;

    @SuppressWarnings("unchecked")
    CircularBuffer(final int size) {
        if (size <= 0) {
            throw new IllegalArgumentException("Buffer size must be positive");
        }
        // Safer array creation
        this.buffer = (T[]) java.lang.reflect.Array.newInstance(Object.class, size);
    }

    T read() throws BufferIOException {
        if (count == 0) {
            throw new BufferIOException("Tried to read from empty buffer");
        }
        T data = buffer[head];
        buffer[head] = null; // Clear the slot
        head = (head + 1) % buffer.length;
        count--;
        return data;
    }

    void write(T data) throws BufferIOException {
        if (data == null) {
            throw new IllegalArgumentException("Cannot write null to buffer");
        }
        if (count == buffer.length) {
            throw new BufferIOException("Tried to write to full buffer");
        }
        buffer[tail] = data;
        tail = (tail + 1) % buffer.length;
        count++;
    }

    void overwrite(T data) {
        if (data == null) {
            throw new IllegalArgumentException("Cannot write null to buffer");
        }
        if (count < buffer.length) {
            write(data);
        } else {
            buffer[head] = data;
            head = (head + 1) % buffer.length;
            tail = (tail + 1) % buffer.length;
        }
    }

    void clear() {
        for (int i = 0; i < buffer.length; i++) {
            buffer[i] = null;
        }
        head = 0;
        tail = 0;
        count = 0;
    }
}
